File: frontend mobile/lib/services/loan_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../models/loan.dart';

class LoanService {
  static const String baseUrl = 'http://10.0.2.2:8080/api/loans';
  final FlutterSecureStorage _secureStorage = FlutterSecureStorage();

  // Get loan by composite ID
  Future<Loan> getLoanById(LoanId id) async {
    try {
      String? token = await _secureStorage.read(key: 'token');
      
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final headers = {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      };

      final response = await http.get(
        Uri.parse('$baseUrl/${id.userId}_${id.bookId}'),
        headers: headers,
      );

      print('Get Loan by ID Response Status: ${response.statusCode}');
      print('Get Loan by ID Response Body: ${response.body}');

      if (response.statusCode == 200) {
        return Loan.fromJson(jsonDecode(response.body));
      } else if (response.statusCode == 404) {
        throw Exception('Loan not found');
      } else {
        throw Exception('Failed to fetch loan. Status code: ${response.statusCode}, Reason: ${response.reasonPhrase}');
      }
    } catch (e) {
      print('Exception in getLoanById: $e');
      throw Exception('Error fetching loan by ID: $e');
    }
  }

  // Get all loans (Admin only)
  Future<List<Loan>> getAllLoans() async {
    try {
      String? token = await _secureStorage.read(key: 'token');
      
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final headers = {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      };

      final response = await http.get(
        Uri.parse(baseUrl),
        headers: headers,
      );

      print('Get All Loans Response Status: ${response.statusCode}');
      print('Get All Loans Response Body: ${response.body}');

      if (response.statusCode == 200) {
        final List<dynamic> loansJson = jsonDecode(response.body);
        return loansJson.map((json) => Loan.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load loans. Status code: ${response.statusCode}, Reason: ${response.reasonPhrase}');
      }
    } catch (e) {
      print('Exception in getAllLoans: $e');
      throw Exception('Error fetching loans: $e');
    }
  }

  // Get loans by user ID
  Future<List<Loan>> getLoansByUserId(int userId) async {
    try {
      String? token = await _secureStorage.read(key: 'token');
      
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final headers = {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      };

      final response = await http.get(
        Uri.parse('$baseUrl/user/$userId'),
        headers: headers,
      );

      print('Get Loans by User ID Response Status: ${response.statusCode}');
      print('Get Loans by User ID Response Body: ${response.body}');

      if (response.statusCode == 200) {
        final List<dynamic> loansJson = jsonDecode(response.body);
        return loansJson.map((json) => Loan.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load user loans. Status code: ${response.statusCode}, Reason: ${response.reasonPhrase}');
      }
    } catch (e) {
      print('Exception in getLoansByUserId: $e');
      throw Exception('Error fetching user loans: $e');
    }
  }

  // Create a new loan
  Future<Loan> createLoan(Loan loan) async {
    try {
      String? token = await _secureStorage.read(key: 'token');
      
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final headers = {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      };

      final response = await http.post(
        Uri.parse(baseUrl),
        headers: headers,
        body: jsonEncode(loan.toJson()),
      );

      print('Create Loan Response Status: ${response.statusCode}');
      print('Create Loan Response Body: ${response.body}');

      if (response.statusCode == 201) {
        return Loan.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to create loan. Status code: ${response.statusCode}, Reason: ${response.reasonPhrase}');
      }
    } catch (e) {
      print('Exception in createLoan: $e');
      throw Exception('Error creating loan: $e');
    }
  }

// Delete a loan (Admin or User based on authorization)
Future<void> deleteLoan(LoanId id) async {
  try {
    String? token = await _secureStorage.read(key: 'token');
    
    if (token == null) {
      throw Exception('No authentication token found');
    }

    final headers = {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };

    final response = await http.delete(
      Uri.parse('$baseUrl/${id.userId}/${id.bookId}'), // Update URL structure
      headers: headers,
    );

    print('Delete Loan Response Status: ${response.statusCode}');

    if (response.statusCode != 204) {
      throw Exception('Failed to delete loan. Status code: ${response.statusCode}, Reason: ${response.reasonPhrase}');
    }
  } catch (e) {
    print('Exception in deleteLoan: $e');
    throw Exception('Error deleting loan: $e');
  }
}
}

File: frontend Angular/src/app/services/loan.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';

// Define the Loan and LoanId interfaces for the payload
export interface Loan {
  id?: {
    userId: number;
    bookId: number;
  }; // Composite primary key
  userId: number;
  bookId: number;
  loanDate: string;
  returnDate: string;
  isReturned: boolean;
  photo?: string; // Book photo
}

export interface LoanId {
  userId: number;
  bookId: number;
}

@Injectable({
  providedIn: 'root',
})
export class LoanService {
  private apiUrl = 'http://localhost:8080/api/loans'; // API endpoint

  constructor(private http: HttpClient) {}

  // Helper to get Authorization headers
  private getAuthHeaders(): HttpHeaders {
    const token = localStorage.getItem('jwtToken');
    console.log('JWT Token:', token); // Debugging
    if (!token) {
      console.error('No JWT token found. Please log in.');
      return new HttpHeaders(); // Return empty headers if token is missing
    }
    return new HttpHeaders({ Authorization: 'Bearer ' + token });
  }

  // Create a new loan
  createLoan(loan: Loan): Observable<Loan> {
    const headers = this.getAuthHeaders();
    const formattedLoan = {
      user: { id: loan.userId }, // Wrap userId inside an object
      book: { id: loan.bookId }, // Wrap bookId inside an object
      loanDate: loan.loanDate,
      returnDate: loan.returnDate,
      isReturned: loan.isReturned,
    };
    return this.http.post<Loan>(`${this.apiUrl}/create`, formattedLoan, { headers }).pipe(
      catchError((error) => {
        console.error('Error creating loan:', error);
        return throwError(error);
      })
    );
  }

  // Check if a book is already borrowed by a user
  checkIfBookAlreadyBorrowed(userId: number, bookId: number): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/check/${userId}/${bookId}`, {
      headers: this.getAuthHeaders(),
    }).pipe(
      catchError((error) => {
        console.error('Error checking if book is borrowed:', error);
        return throwError(error);
      })
    );
  }

  // Get loans by user ID
  getLoansByUserId(userId: number): Observable<Loan[]> {
    return this.http.get<Loan[]>(`${this.apiUrl}/user/${userId}`, {
      headers: this.getAuthHeaders(),
    }).pipe(
      catchError((error) => {
        console.error('Error fetching loans:', error);
        return throwError(error);
      })
    );
  }

  // Get loan history by user ID
  getLoanHistoryByUserId(userId: number): Observable<Loan[]> {
    return this.http.get<Loan[]>(`${this.apiUrl}/user/${userId}/history`, {
      headers: this.getAuthHeaders(),
    }).pipe(
      catchError((error) => {
        console.error('Error fetching loan history:', error);
        return throwError(error);
      })
    );
  }

  // Update loan status (e.g., mark as returned

  // Delete a loan by LoanId
  deleteLoan(loanId: LoanId): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${loanId.userId}/${loanId.bookId}`, {
      headers: this.getAuthHeaders(),
    }).pipe(
      catchError((error) => {
        console.error('Error deleting loan:', error);
        return throwError(error);
      })
    );
  }

  // Get book details
  getBookDetails(bookId: number): Observable<{ id: number; title: string }> {
    const url = `http://localhost:8080/api/books/${bookId}`;
    return this.http.get<{ id: number; title: string }>(url, {
      headers: this.getAuthHeaders(),
    }).pipe(
      catchError((error) => {
        console.error(`Error fetching book details for bookId ${bookId}:`, error);
        return throwError(error);
      })
    );
  }

  // Fetch all loans
getAllLoans(): Observable<Loan[]> {
  return this.http.get<Loan[]>(`${this.apiUrl}`, {
    headers: this.getAuthHeaders(),
  }).pipe(
    catchError((error) => {
      console.error('Error fetching all loans:', error);
      return throwError(error);
    })
  );
}

updateLoanStatuses(): Observable<any> {
  const headers = this.getAuthHeaders();
  return this.http.post(`${this.apiUrl}/update-loans-status`, {}, { headers }).pipe(
    catchError((error) => {
      console.error('Error updating loan statuses:', error);
      return throwError(error);
    })
  );
}


}
